source/fuzz/transformation_duplicate_region_with_selection.cpp

```diff
@@ -0,0 +1,566 @@
+    if (dominator_analysis->Dominates(entry_block, &block) &&
+        postdominator_analysis->Dominates(exit_block, &block)) {
+      result.insert(&block);
+    }
+  }
+  return result;
+}
+
+void TransformationDuplicateRegionWithSelection::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  fuzzerutil::UpdateModuleIdBound(ir_context, message_.new_entry_fresh_id());
+  fuzzerutil::UpdateModuleIdBound(ir_context, message_.merge_label_fresh_id());
+
+  // Create the new entry block containing the main conditional instruction. Set
+  // its parent to the parent of the original entry block, since it is located
+  // in the same function.
+  std::unique_ptr<opt::BasicBlock> new_entry_block =
+      MakeUnique<opt::BasicBlock>(MakeUnique<opt::Instruction>(
+          ir_context, SpvOpLabel, 0, message_.new_entry_fresh_id(),
+          opt::Instruction::OperandList()));
+  auto entry_block = ir_context->cfg()->block(message_.entry_block_id());
+  auto enclosing_function = entry_block->GetParent();
+  auto exit_block = ir_context->cfg()->block(message_.exit_block_id());
+
+  // Get the blocks contained in the region.
+  std::set<opt::BasicBlock*> region_blocks =
+      GetRegionBlocks(ir_context, entry_block, exit_block);
+
+  // Construct the merge block and set its parent.
+  std::unique_ptr<opt::BasicBlock> merge_block =
+      MakeUnique<opt::BasicBlock>(MakeUnique<opt::Instruction>(opt::Instruction(
+          ir_context, SpvOpLabel, 0, message_.merge_label_fresh_id(),
+          opt::Instruction::OperandList())));
+
+  // Get the maps from the protobuf.
+  std::map<uint32_t, uint32_t> original_label_to_duplicate_label =
+      fuzzerutil::RepeatedUInt32PairToMap(
+          message_.original_label_to_duplicate_label());
+
+  std::map<uint32_t, uint32_t> original_id_to_duplicate_id =
+      fuzzerutil::RepeatedUInt32PairToMap(
+          message_.original_id_to_duplicate_id());
+
+  std::map<uint32_t, uint32_t> original_id_to_phi_id =
+      fuzzerutil::RepeatedUInt32PairToMap(message_.original_id_to_phi_id());
+
+  // Before adding duplicate blocks, we need to update the OpPhi instructions in
+  // the successors of the |exit_block|. We know that the execution of the
+  // transformed region will end in |merge_block|. Hence, we need to change all
+  // occurrences of the label id of the |exit_block| to the label id of the
+  // |merge_block|.
+  exit_block->ForEachSuccessorLabel([this, ir_context](uint32_t label_id) {
+    auto block = ir_context->cfg()->block(label_id);
+    for (auto& instr : *block) {
+      if (instr.opcode() == SpvOpPhi) {
+        instr.ForEachId([this](uint32_t* id) {
+          if (*id == message_.exit_block_id()) {
+            *id = message_.merge_label_fresh_id();
+          }
+        });
+      }
+    }
+  });
+
+  // Similarly, we need to update OpPhi instructions in the |entry_block|. We
+  // know that it has only one predecessor, since the region is single-entry,
```

Comment:
The entry block of a single-entry-single-exit region can have more than one predecessor.  "Single-entry" just means that there is only one way to get into the region - i.e., via the entry block.  It does not mean that the entry block is restricted to having just one predecessor.

Please write a test case for duplicating a region with a selection for the case where the region entry block has two predecessors, e.g.

```
     A
    / \
   B   C
    \ /
   ENTRY
     |
    EXIT
     |
     D
```

and add it to the test file, fixing up whatever is necessary to ensure that it works.

Reply:
Note that we require that if a construct header is in the region then it's merge is also in the region.  But we allow the case where the merge is in the region but the header is not.

Reply:
I've checked it and the current code doesn't allow a situation where the loop header is not in the region, but the merge block is. I don't think that there is a possibility of entry block having more than one predecessor.

Reply:
Thanks for checking!  Can you please add an assertion to confirm that there is only one predecessor?

Reply:
Done. Thanks.